### PR TITLE
Add TypeScript type declaration file (fixes #148)

### LINKED ISSUE #148: Add @types/fast-json-stringify for us TypeScript users
Always want types with my TypeScript. 

Happy to help out here.

### PR DESCRIPTION
**Pull request by mjwwit**
This wasn't my issue, but in order to get to know the module I found myself writing TypeScript types for it. It should fix #148.

I've made an attempt to also add types for the subset of the json-schema standard this module uses. This should help TypeScript users to stay away from unsupported features.

### CODE REVIEW

**Comment by mcollina** [comment]
Awesome work! Would you mind adding some tests like Fastify does? https://github.com/fastify/fastify/tree/master/test/types

**Comment by mcollina** [comment]
Awesome work! Would you mind adding some tests like Fastify does? https://github.com/fastify/fastify/tree/master/test/types

**Comment by Ethan-Arrowood** [comment]
I'm excited to look over this. I'd recommend trying to test the types using this new library [`tsd`](https://github.com/SamVerschueren/tsd#readme). You can see how I'm using it for `fastify/light-my-request` in this [pr](https://github.com/fastify/light-my-request/pull/39)

**Comment by Ethan-Arrowood** [comment]
I'd like to see some implementation examples before I start recommending changes. I'd recommend adding it to the README (or through the tsd test file I mentioned in my last comment)

File: fast-json-stringify.d.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
+declare module "fast-json-stringify" {
```
Comment (Ethan-Arrowood) [inline fast-json-stringify.d.ts]: Remove this declaration. Declare the `build` namespace on the top-level i.e. `declare namespace build { ... }`

File: fast-json-stringify.d.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+      | ArraySchema
+      | TupleSchema
+      | ObjectSchema;
+  }
+
+  /**
+   * Build a stringify function using a schema of the documents that should be stringified
+   * @param schema The schema used to stringify values
+   */
+  function build(schema: build.StringSchema): (doc: string) => string;
```
Comment (Ethan-Arrowood) [inline fast-json-stringify.d.ts]: add `declare` keyword to each of these function declarations so they are all exported i.e. `declare function build(...) ...`

File: fast-json-stringify.d.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  function build(schema: build.IntegerSchema | build.NumberSchema): (doc: number) => string;
+  function build(schema: build.NullSchema): (doc: null) => 'null';
+  function build(schema: build.BooleanSchema): (doc: boolean) => string;
+  function build(schema: build.ArraySchema | build.TupleSchema): (doc: any[]) => string;
+  function build(schema: build.ObjectSchema): (doc: object) => string;
+  function build(schema: build.Schema): (doc: object | any[] | string | number | boolean | null) => string;
+  export = build;
+}
```
Comment (Ethan-Arrowood) [inline fast-json-stringify.d.ts]: When removing the `module` declaration don't forget to remove this `}` character. 

The `build` namespace and function declarations should live at the top-level of this file along with the `export = build` line.

File: fast-json-stringify.d.ts [hunk 4]
```diff
@@ -0,0 +1,143 @@
+declare module "fast-json-stringify" {
+  namespace build {
```
Comment (Ethan-Arrowood) [inline fast-json-stringify.d.ts]: Change this to `declare namespace build { ... }`

**Comment by mjwwit** [comment]
> I'm excited to look over this. I'd recommend trying to test the types using this new library [`tsd`](https://github.com/SamVerschueren/tsd#readme). You can see how I'm using it for `fastify/light-my-request` in this [pr](https://github.com/fastify/light-my-request/pull/39)

I somehow missed this comment and implemented tests like in the fastify project. I could still look into this if it's important enough, let me know.

**Comment by Ethan-Arrowood** [comment]
I would like to see testing with `tsd` however I will approve these changes so we can get your types merged and published. We can track tsd testing in a separate issue

**Comment by Ethan-Arrowood** [comment]
I would like to see testing with `tsd` however I will approve these changes so we can get your types merged and published. We can track tsd testing in a separate issue

**Comment by Ethan-Arrowood** [comment]
@delvedor @mcollina I don't have permissions to publish this module to npm. I have bumped the minor version and pushed the tag to GitHub (as well as the updated package.json). I guess either one of you will have to `npm publish` or if you grant me permissions I can do it :+1:

**Comment by mjwwit** [comment]
@Ethan-Arrowood I don't think your push of the tag worked, as I can't see a v1.14 tag...

**Comment by mcollina** [comment]
released

**Comment by Ethan-Arrowood** [comment]
Tag pushed too!

**Pull request merged**